stores.js

```diff
@@ -0,0 +1,260 @@
+    if (o == null) return o
+    return Mutable.decode(o)
+  }
+}
+
+// PUT_VALUE_MAX_SIZE + packet overhead (i.e. the key etc.)
+// should be less than the network MTU, normally 1400 bytes
+const PUT_VALUE_MAX_SIZE = 1000
+
+class ImmutableStore {
+  constructor (dht, store) {
+    this.dht = dht
+    this.store = store
+  }
+  get (key, cb) {
+    if (!Buffer.isBuffer(key)) throw Error('Key must be a buffer')
+    const { store, dht } = this
+    // if the querying node already has the immutable value
+    // then there's no need to query the dht
+    const streamMode = typeof cb !== 'function'
+    const hexKey = key.toString('hex')
+    const value = store.get(hexKey)
+    if (value && streamMode === false) {
+      return process.nextTick(cb, null, value)
+    }
+
+    let found = false
+    const queryStream = dht.query('immutable-store', key)
+
+    const responseStream = Transform({
+      objectMode: true,
+      transform (result, enc, next) {
+        if (result.value === null) {
+          next()
+          return
+        }
+        const check = Buffer.alloc(32)
+        hash(check, result.value)
+        if (Buffer.compare(check, key) === 0) {
+          if (streamMode) {
+            next(null, result.value)
+          } else {
+            found = true
+            cb(null, result.value)
+            queryStream.destroy()
+          }
+        }
+      }
+    })
+    if (value && streamMode) {
+      // push local cached value to stream first
+      responseStream.push(value)
+    }
+    pipeline(queryStream, responseStream, (err) => {
+      if (err) {
+        cb(err)
+        return
+      }
+      if (streamMode === false && found === false) {
+        cb(null, null)
+      }
+    })
+    if (streamMode) return responseStream
+  }
+  put (value, cb) {
+    if (typeof cb !== 'function') throw Error('Callback is required')
+    if (!Buffer.isBuffer(value)) throw Error('Value must be a buffer')
+    if (value.length > PUT_VALUE_MAX_SIZE) { throw Error(`Value size must be <= ${PUT_VALUE_MAX_SIZE}`) }
+    const { store, dht } = this
+    const key = Buffer.alloc(32)
+    hash(key, value)
+    // set locally for easy cached retrieval
+    store.set(key.toString('hex'), value)
+    // send to the dht
+    dht.update('immutable-store', key, value, (err) => {
+      if (err) {
+        cb(err)
+        return
+      }
+      cb(null, key)
+    })
+  }
+  _command () {
+    const { store } = this
+    return {
+      update ({ target, value }, cb) {
+        const key = Buffer.alloc(32)
+        hash(key, value)
+        if (Buffer.compare(key, target) !== 0) {
+          cb(Error('ERR_INVALID_INPUT'))
+          return
+        }
+        store.set(key.toString('hex'), value)
+        cb(null)
+      },
+      query ({ target }, cb) {
+        cb(null, store.get(target.toString('hex')))
+      }
+    }
+  }
+}
+class MutableStore {
+  constructor (dht, store) {
+    this.dht = dht
+    this.store = store
+  }
+  salt (size = 16) {
```

Comment:
special reason for salt to 16? otherwise it's safer to default it to 32 bytes unless you know what you are doing

Reply:
only reason was to reduce payload sizes - will change